app/routes/thematic/geoencoder.py

```diff
@@ -0,0 +1,216 @@
+                            if adm_level >= 2
+                            else None
+                        ),
+                        "name": match["name_2"] if adm_level >= 2 else None,
+                    },
+                }
+                for match in json_data
+            ],
+        }
+    )
+
+
+def sanitize_names(
+    normalize_search: bool,
+    country: str | None,
+    region: str | None,
+    subregion: str | None,
+) -> List[str | None]:
+    """Turn any empty strings into Nones, enforces the admin level hierarchy,
+    and optionally unaccents and decapitalizes names.
+    """
+    names = []
+
+    if subregion and not region:
+        raise HTTPException(
+            status_code=400,
+            detail="If subregion is specified, region must be specified as well.",
+        )
+
+    for name in (country, region, subregion):
+        if name and normalize_search:
+            names.append(unidecode(name).lower())
+        elif name:
+            names.append(name)
+        else:
+            names.append(None)
+    return names
+
+
+def determine_admin_level(
+    country: str | None, region: str | None, subregion: str | None
+) -> int:
+    """Infer the native admin level of a request based on the presence of
+    non-empty fields
+    """
+    if subregion:
+        return 2
+    elif region:
+        return 1
+    elif country:
+        return 0
+    else:  # Shouldn't get here if FastAPI route definition worked
+        raise HTTPException(status_code=400, detail="Country MUST be specified.")
+
+
+def _admin_boundary_lookup_sql(
+    adm_level: int,
+    normalize_search: bool,
+    dataset: str,
+    country_name: str,
+    region_name: str | None,
+    subregion_name: str | None,
+) -> str:
+    """Generate the SQL required to look up administrative boundary
+    IDs by name.
+    """
+    name_fields: List[str] = ["country", "name_1", "name_2"]
+    if normalize_search:
+        match_name_fields = [name_field + "_normalized" for name_field in name_fields]
+    else:
+        match_name_fields = name_fields
+
+    sql = (
+        f"SELECT gid_0, gid_1, gid_2, {name_fields[0]}, {name_fields[1]}, {name_fields[2]}"
+        f" FROM {dataset} WHERE {match_name_fields[0]}='{country_name}'"
+    )
+    if region_name is not None:
+        sql += f" AND {match_name_fields[1]}='{region_name}'"
+    if subregion_name is not None:
+        sql += f" AND {match_name_fields[2]}='{subregion_name}'"
+
+    sql += f" AND adm_level='{adm_level}'"
+
+    return sql
+
+
+async def version_is_valid(
+    dataset: str,
+    version: str,
+) -> None:
+    """Validate a version string for a given dataset."""
+    # Note: At some point I intend to change the version validator to
+    # use messaging like this. However, that is out of scope for the
+    # current ticket, and I want to see what people think, so I'll
+    # keep it here for now.
+    if re.match(VERSION_REGEX, version) is None:
```

Comment:
per above, maybe just have pre-validated versions rather than checking the version

Reply:
Now specified on a per-env basis!